staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -44,8 +49,88 @@ const (
 	secretboxTransformerPrefixV1 = "k8s:enc:secretbox:v1:"
 	kmsTransformerPrefixV1       = "k8s:enc:kms:v1:"
 	kmsPluginConnectionTimeout   = 3 * time.Second
+	kmsPluginHealthzRate         = rate.Limit(1)
+	kmsPluginHealthzBurst        = 1
 )
 
+// KMSPluginProbe checks health status of KMS-Plugins.
+type KMSPluginProbe struct {
+	name string
+	envelope.Service
+	*rate.Limiter
+}
+
+// GetKMSPluginProbes extracts KMSPluginProbes from the EncryptionConfig.
+func GetKMSPluginProbes(filepath string) ([]*KMSPluginProbe, error) {
+	f, err := os.Open(filepath)
+	if err != nil {
+		return nil, fmt.Errorf("error opening encryption provider configuration file %q: %v", filepath, err)
+	}
+	defer f.Close()
+	return getKMSPluginProbes(f)
+}
+
+func getKMSPluginProbes(reader io.Reader) ([]*KMSPluginProbe, error) {
+	var result []*KMSPluginProbe
+
+	configFileContents, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return result, fmt.Errorf("could not read content of encryption provider configuration: %v", err)
+	}
+
+	config, err := loadConfig(configFileContents)
+	if err != nil {
+		return result, fmt.Errorf("error while parsing encrypiton provider configuration: %v", err)
+	}
+
+	for _, r := range config.Resources {
+		for _, p := range r.Providers {
+			if p.KMS != nil && p.KMS.InstallHealthz {
+				timeout := kmsPluginConnectionTimeout
+				if p.KMS.Timeout != nil {
+					if p.KMS.Timeout.Duration <= 0 {
+						return nil, fmt.Errorf("could not configure KMS-Plugin's probe %q, timeout should be a positive value", p.KMS.Name)
+					}
+					timeout = p.KMS.Timeout.Duration
+				}
+
+				s, err := envelope.NewGRPCService(p.KMS.Endpoint, timeout)
+				if err != nil {
+					return nil, fmt.Errorf("could not configure KMS-Plugin's probe %q, error: %v", p.KMS.Name, err)
+				}
+				result = append(result, &KMSPluginProbe{
+					name:    p.KMS.Name,
+					Service: s,
+					Limiter: rate.NewLimiter(kmsPluginHealthzRate, kmsPluginHealthzBurst),
+				})
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// Check encrypts and decrypts test data against KMS-Plugin's gRPC endpoint.
+func (h *KMSPluginProbe) Check() error {
+	if err := h.Limiter.Wait(context.Background()); err != nil {
```

Comment:
This would delay the health check.  Don't you just want to remember the last status and if you cannot immediately run the check return the last known status?

Reply:
Done.